ironfish/src/wallet/queue.ts

```diff
@@ -0,0 +1,52 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+export class Queue<T> {
+  private _items: T[] = []
+
+  constructor(private capacity: number = Infinity) {}
+
+  /**
+   * Adds @param item to the queue as the last element
+   */
+  enqueue(item: T): void {
+    if (this.isFull()) {
+      throw Error('Queue has reached max capacity, you cannot add more items')
+    }
+    this._items.push(item)
+  }
+
+  /**
+   * Remove the first element in the queue. If the queue is empty, returns `undefined`
+   */
+  dequeue(): T | undefined {
+    return this._items.shift()
+  }
+
+  /**
+   * return the number of elements in the queue
+   */
+  size(): number {
+    return this._items.length
+  }
+
+  /**
+   * Returns true if the queue size reaches the capacity
+   */
+  isFull(): boolean {
+    return this.capacity === this.size()
+  }
+
+  /**
+   * Get the item in the queue by index
+   */
+  get(index: number): T | undefined {
+    if (index < 0 || index >= this._items.length) {
+      return undefined
+    }
+    return this._items[index]
+  }
+}
```

Comment:
The built in javascript array is a queue, so we can just do that.

Reply:
This is queue with size, not just an array. There are enqueue logic that encapsulate in the queue class.

Reply:
There is some logic around the maximum capacity, but since `enqueue` throws an error if the queue is at capacity the client still needs to keep track of whether the queue is full.

I'm not sure if the extra abstraction is needed here

Reply:
I did not include automatically dequeue when queue is full during enqueuing. I left it for the caller to handle what to do when queue is full.

The queue class store the capacity so it knows when the queue is full. If we have the capacity at recent fee cache class is also doable. But I think this way the recent fee cache class is cleaner.